### PR TITLE
C#: Better call analysis using CIL

### PR DESCRIPTION
**Pull request by calumgrant**
Various improvements to the CIL analysis :-

- Identifying assemblies that are stubs. This is needed to exclude stub code from functions that always return null or always throw exceptions. Various approaches were attempted, in order to find something simple, robust and efficient. For example, identifying `throw null` or the proportion of functions that returned `null`. In the end, using the average method size (number of instructions) proved to be a reliable metric.
- Identify the best/unique implementation for a CIL method, using `CIL::Method.getImplementation()`, which should make dataflow more efficient by only considering at most one implementation.
- Fix `FalseFlow` edges which always pointed to the `TrueFlow` edges. This didn't affect anything because no queries relied on this behaviour.
- Identify CIL methods that always throw exceptions, and incorporate them into the C# control flow graph.
- Identify CIL methods that always return null or non-null values, to make the C# nullness queries more reliable.
- Fix expressions of type `default(object)` which were incorrectly `NonNullExpr`.

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/CallableReturns.qll [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/**
+ * Provides predicates for analysing the return values of callables.
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/dataflow/internal/CallableReturns.qll]: analyzing

File: csharp/ql/src/semmle/code/csharp/dataflow/internal/CallableReturns.qll [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  or
+  exists(CIL::Method m, CIL::Type t | m.matchesHandle(c) | CR::alwaysThrowsException(m, t) and t.matchesHandle(ex))
+}
+
+/** Holds if callable `m' can return expression `ret` directly or indirectly. */
+private predicate callableReturns(Callable m, Expr e) {
+  exists(DataFlow::Node srcNode, DataFlow::Node retNode |
+    e = srcNode.asExpr() and m.canReturn(retNode.asExpr())
+  |
+    DataFlow::localFlow(srcNode, retNode)
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/dataflow/internal/CallableReturns.qll]: The same comments regarding local flow as for CIL apply here.

Note that `NonNullExpr` already [takes local flow into account](https://github.com/Semmle/ql/blob/master/csharp/ql/src/semmle/code/csharp/dataflow/Nullness.qll#L72), so we should probably build something similar into [`AlwaysNullExpr`](https://github.com/Semmle/ql/blob/master/csharp/ql/src/semmle/code/csharp/dataflow/Nullness.qll#L48).

File: change-notes/1.20/analysis-csharp.md [hunk 3]
```diff
@@ -18,6 +18,7 @@
 | Use of default ToString() (cs/call-to-object-tostring) | Fewer false positives | Results have been removed for `char` arrays passed to `StringBuilder.Append()`, which were incorrectly marked as using `ToString`. |
 | Use of default ToString() (cs/call-to-object-tostring) | Fewer results | Results have been removed when the object is an interface or an abstract class. |
 | Unused format argument (cs/format-argument-unused) | Fewer false positives | Results have been removed where the format string is empty. This is often used as a default value and is not an interesting result. |
+| Dereferenced variable may be null (cs/dereferenced-value-may-be-null), Dereferenced variable is always null (cs/dereferenced-value-is-always-null) | Improved results | Use CIL analysis to find calls that always return null, or non-null values. |
```
Comment (jf205) [inline change-notes/1.20/analysis-csharp.md]: Format query id as code.

**Comment by calumgrant** [comment]
I think this is ready for review now. The main changes are:

- Improve the dataflow tests by creating a test DLL. Unfortunately, the generation of this DLL is still a manual process, but is much better than just testing the standard library.
- Fix various bugs discovered during testing.

Performance report is [here](https://git.semmle.com/gist/calum/54de2871f0d50165f659f48f1eb00513). Overall, the performance is fine, and changes to the results are positive, although there is also some instability as previously observed elsewhere.

**Comment by hvitved** [comment]
I ran a [dist comparison](https://git.semmle.com/gist/tom/d13781fe5b20eb740f2183ed2a5a48f3) (internal link), and with the exception of `monodevelop`, things look good. However, the slowdown on `monodevelop` is not enough that I think it should block this PR, instead I can tak a look and see if a follow-up PR is needed.

**Pull request merged**